### ISSUE TITLE
Triton failed to open the cudaIpcHandle upon prediction request when launched in container under WSL2

### ISSUE THREAD

**Issue by BorisPolonsky** [triton-inference-server/server#4915]
**Description**
Triton failed to open cudaIpcHandle while trying to serve model exported from official NVIDIA/Merlin [notebook showcase](https://github.com/NVIDIA-Merlin/systems/blob/v0.5.0/examples/Serving-An-XGboost-Model-With-Merlin-Systems.ipynb), where user will train and export an ensembled model, serve it with triton, make prediction via gRPC calls.

**Triton Information**
What version of Triton are you using?
Triton 22.03.0, the one that ships with `nvcr.io/nvidia/merlin/merlin-tensorflow:22.07`
Are you using the Triton container or did you build it yourself?
Container. Launched under WSL2
**To Reproduce**
Steps to reproduce the behavior.
- Follow [this comment](https://github.com/triton-inference-server/server/issues/4915#issuecomment-1258910783) in this issue, which does the following things
  - Export model refering to this [issue](https://github.com/NVIDIA-Merlin/systems/issues/205), or alternatively, download [here](https://drive.google.com/drive/folders/15Slql2i8jncYB6H1Gxkr1q4jPPvxG9Xk?usp=sharing).
  - Serve it in triton `docker run -it --name merlin -v /home:/home -e LANG=C.UTF-8 -v /home:/home --gpus all --ipc=host --ulimit memlock=-1 --ulimit stack=67108864 -p 6006:6006 nvcr.io/nvidia/merlin/merlin-tensorflow:22.07 bash`
  - Send predcition request with gRPC calls. Use the code block below "Let's now package the information up as inputs and send it to Triton for inference." in the [notebook](https://github.com/NVIDIA-Merlin/systems/blob/v0.5.0/examples/Serving-An-XGboost-Model-With-Merlin-Systems.ipynb) mentioned above.

And triton failed to process the request for `1_predictforest`
```
0923 05:19:56.350606 2129 pb_stub.cc:776] Failed to process the request(s) for model '1_predictforest', message: TypeError: __init__(): incompatible constructor arguments. The following argument types are supported:
    1. c_python_backend_utils.InferenceResponse(output_tensors: List[c_python_backend_utils.Tensor], error: c_python_backend_utils.TritonError = None)

Invoked with: kwargs: tensors=[], error="<class 'c_python_backend_utils.TritonModelException'>, Failed to open the cudaIpcHandle. error: unknown error, [<FrameSummary file /home/polonsky/Documents/merlin-systems-0.5.0/examples/ensemble/1_predictforest/1/model.py, line 105 in execute>, <FrameSummary file /usr/local/lib/python3.8/dist-packages/merlin/systems/dag/op_runner.py, line 38 in execute>, <FrameSummary file /usr/local/lib/python3.8/dist-packages/merlin/systems/dag/ops/fil.py, line 160 in transform>]"

At:
  /home/polonsky/Documents/merlin-systems-0.5.0/examples/ensemble/1_predictforest/1/model.py(122): execute
```
Describe the models (framework, inputs, outputs), ideally include the model configuration file (if using an ensemble include the model configuration file for that as well).
An ensenmbled XGBoost model
```
name: "ensemble_model"
platform: "ensemble"
input {
  name: "movieId"
  data_type: TYPE_INT32
  dims: -1
  dims: -1
}
input {
  name: "userId"
  data_type: TYPE_INT32
  dims: -1
  dims: -1
}
input {
  name: "genres"
  data_type: TYPE_INT32
  dims: -1
  dims: -1
}
input {
  name: "TE_movieId_rating"
  data_type: TYPE_FP64
  dims: -1
  dims: -1
}
input {
  name: "userId_count"
  data_type: TYPE_FP32
  dims: -1
  dims: -1
}
input {
  name: "gender"
  data_type: TYPE_INT32
  dims: -1
  dims: -1
}
input {
  name: "zip_code"
  data_type: TYPE_INT32
  dims: -1
  dims: -1
}
input {
  name: "age"
  data_type: TYPE_INT32
  dims: -1
  dims: -1
}
output {
  name: "output__0"
  data_type: TYPE_FP32
  dims: -1
  dims: -1
}
ensemble_scheduling {
  step {
    model_name: "0_transformworkflow"
    model_version: -1
    input_map {
      key: "TE_movieId_rating"
      value: "TE_movieId_rating"
    }
    input_map {
      key: "age"
      value: "age"
    }
    input_map {
      key: "gender"
      value: "gender"
    }
    input_map {
      key: "genres"
      value: "genres"
    }
    input_map {
      key: "movieId"
      value: "movieId"
    }
    input_map {
      key: "userId"
      value: "userId"
    }
    input_map {
      key: "userId_count"
      value: "userId_count"
    }
    input_map {
      key: "zip_code"
      value: "zip_code"
    }
    output_map {
      key: "TE_movieId_rating"
      value: "TE_movieId_rating_0"
    }
    output_map {
      key: "age"
      value: "age_0"
    }
    output_map {
      key: "gender"
      value: "gender_0"
    }
    output_map {
      key: "genres"
      value: "genres_0"
    }
    output_map {
      key: "movieId"
      value: "movieId_0"
    }
    output_map {
      key: "userId"
      value: "userId_0"
    }
    output_map {
      key: "userId_count"
      value: "userId_count_0"
    }
    output_map {
      key: "zip_code"
      value: "zip_code_0"
    }
  }
  step {
    model_name: "1_predictforest"
    model_version: -1
    input_map {
      key: "TE_movieId_rating"
      value: "TE_movieId_rating_0"
    }
    input_map {
      key: "age"
      value: "age_0"
    }
    input_map {
      key: "gender"
      value: "gender_0"
    }
    input_map {
      key: "genres"
      value: "genres_0"
    }
    input_map {
      key: "movieId"
      value: "movieId_0"
    }
    input_map {
      key: "userId"
      value: "userId_0"
    }
    input_map {
      key: "userId_count"
      value: "userId_count_0"
    }
    input_map {
      key: "zip_code"
      value: "zip_code_0"
    }
    output_map {
      key: "output__0"
      value: "output__0"
    }
  }
}
```
And here's what `1_predictforest/1/model.py` look like:
```
# Copyright (c) 2022, NVIDIA CORPORATION. All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#  * Neither the name of NVIDIA CORPORATION nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

import json
import pathlib
import sys
import traceback

import triton_python_backend_utils as pb_utils

from merlin.systems.dag.op_runner import OperatorRunner
from merlin.systems.dag.ops.operator import InferenceDataFrame


class TritonPythonModel:
    """Model for Triton Python Backend.

    Every Python model must have "TritonPythonModel" as the class name
    """

    def initialize(self, args):
        """Called only once when the model is being loaded. Allowing
        the model to initialize any state associated with this model.

        Parameters
        ----------
        args : dict
          Both keys and values are strings. The dictionary keys and values are:
          * model_config: A JSON string containing the model configuration
          * model_instance_kind: A string containing model instance kind
          * model_instance_device_id: A string containing model instance device ID
          * model_repository: Model repository path
          * model_version: Model version
          * model_name: Model name
        """
        self.model_config = json.loads(args["model_config"])

        self.runner = OperatorRunner(
            self.model_config,
            # model_repository=_parse_model_repository(args["model_repository"]),
            # model_name=args["model_name"],
            # model_version=args["model_version"],
        )

    def execute(self, requests):
        """Receives a list of pb_utils.InferenceRequest as the only argument. This
        function is called when an inference is requested for this model. Depending on the
        batching configuration (e.g. Dynamic Batching) used, `requests` may contain
        multiple requests. Every Python model, must create one pb_utils.InferenceResponse
        for every pb_utils.InferenceRequest in `requests`. If there is an error, you can
        set the error argument when creating a pb_utils.InferenceResponse.

        Parameters
        ----------
        requests : list
          A list of pb_utils.InferenceRequest

        Returns
        -------
        list
          A list of pb_utils.InferenceResponse. The length of this list must
          be the same as `requests`
        """
        params = self.model_config["parameters"]
        op_names = json.loads(params["operator_names"]["string_value"])
        first_operator_name = op_names[0]
        operator_params = json.loads(params[first_operator_name]["string_value"])
        input_column_names = list(json.loads(operator_params["input_dict"]).keys())

        responses = []

        for request in requests:
            try:
                # transform the triton tensors to a dict of name:numpy tensor
                input_tensors = {
                    name: pb_utils.get_input_tensor_by_name(request, name).as_numpy()
                    for name in input_column_names
                }

                inf_df = InferenceDataFrame(input_tensors)

                raw_tensor_tuples = self.runner.execute(inf_df)

                output_tensors = []
                for name, data in raw_tensor_tuples:
                    if isinstance(data, pb_utils.Tensor):
                        output_tensors.append(data)
                        continue
                    data = data.get() if hasattr(data, "get") else data
                    tensor = pb_utils.Tensor(name, data)
                    output_tensors.append(tensor)

                responses.append(pb_utils.InferenceResponse(output_tensors))

            except Exception:  # pylint: disable=broad-except
                exc_type, exc_value, exc_traceback = sys.exc_info()
                tb_string = repr(traceback.extract_tb(exc_traceback))
                responses.append(
                    pb_utils.InferenceResponse(
                        tensors=[], error=f"{exc_type}, {exc_value}, {tb_string}"
                    )
                )

        return responses


def _parse_model_repository(model_repository: str) -> str:
    """
    Extract the model repository path from the model_repository value
    passed to the TritonPythonModel initialize method.
    """
    # Handle bug in Tritonserver 22.06
    # model_repository argument became path to model.py
    # instead of path to model directory within the model repository
    if model_repository.endswith(".py"):
        return str(pathlib.Path(model_repository).parent.parent.parent)
    else:
        return str(pathlib.Path(model_repository).parent)

```
**Expected behavior**

Triton inference server should return the desired output properly, unless it's something related to WSL2 that leads to the problem.

**Comment by BorisPolonsky** [comment]
A quicker way to replicate this issue
- Build Image based on `nvcr.io/nvidia/merlin/merlin-tensorflow:22.07`
```
FROM nvcr.io/nvidia/merlin/merlin-tensorflow:22.07
WORKDIR /models
RUN git pull origin main && pip install .
WORKDIR /nvtabular
RUN git pull origin main && pip install .
WORKDIR /core
RUN git pull origin main && pip install .
WORKDIR /systems
RUN git pull origin main && pip install .
WORKDIR /opt/tritonserver
```
- Conduct `pytest`
```
 docker run --rm --workdir /systems -e LANG=C.UTF-8 -v /etc/localtime:/etc/localtime:ro --gpus all
--ipc=host --ulimit memlock=-1 --ulimit stack=67108864 $YOUR_IMAGE pytest tests/unit/examples/test_serving_an_xgboost_model_with_merlin_systems.py

**Comment by tanmayv25** [comment]
@BorisPolonsky  Can you try running the logic in script in `model.py` directly on WSL2 outside Triton and let us know if it works for you?

**Comment by BorisPolonsky** [comment]
The problem is solved in newer version of `triton` (`2.25.0`, the one that ships with `nvcr.io/nvidia/merlin/merlin-tensorflow:22.09`)

I've closed the related [issue](https://github.com/NVIDIA-Merlin/systems/issues/205).

**Comment by youwan114** [comment]
Hello @BorisPolonsky @tanmayv25   I also meet this problem, but I use the version of container is **nvcr.io/nvidia/tritonserver:23.07**，upon wsl2 ubuntu 20.04 on win11. could you mind tell me how you solved this problem?
